bgpd/bgp_bfd.c

```diff
@@ -40,667 +41,392 @@
+	/* Set defaults. */
+	p->bfd_config->detection_multiplier = BFD_DEF_DETECT_MULT;
+	p->bfd_config->min_rx = BFD_DEF_MIN_RX;
+	p->bfd_config->min_tx = BFD_DEF_MIN_TX;
+	p->bfd_config->cbit = false;
+	p->bfd_config->profile[0] = 0;
 }
 
-/**
- * bgp_bfd_deregister_peer - deregister a peer with BFD through zebra
- *                           for stopping the monitoring of the peer
- *                           rechahability.
- */
-void bgp_bfd_deregister_peer(struct peer *peer)
+void bgp_peer_configure_bfd(struct peer *p, bool manual)
 {
-	struct bfd_info *bfd_info;
+	/* Groups should not call this. */
+	assert(!CHECK_FLAG(p->sflags, PEER_STATUS_GROUP));
 
-	if (!peer->bfd_info)
-		return;
-	bfd_info = (struct bfd_info *)peer->bfd_info;
+	/* Already configured, skip it. */
+	if (p->bfd_config) {
+		/* If manually active update flag. */
+		if (!p->bfd_config->manual)
+			p->bfd_config->manual = manual;
 
-	/* Check if BFD is eanbled and peer has not been registered */
-	if (!CHECK_FLAG(bfd_info->flags, BFD_FLAG_BFD_REG))
 		return;
+	}
 
-	bfd_info->status = BFD_STATUS_DOWN;
-	bfd_info->last_update = bgp_clock();
-
-	bgp_bfd_peer_sendmsg(peer, ZEBRA_BFD_DEST_DEREGISTER);
-}
+	/* Allocate memory for configuration overrides. */
+	p->bfd_config = XCALLOC(MTYPE_BFD_CONFIG, sizeof(*p->bfd_config));
+	p->bfd_config->manual = manual;
 
-/*
- * bgp_bfd_update_peer - update peer with BFD with new BFD paramters
- *                       through zebra.
- */
-static void bgp_bfd_update_peer(struct peer *peer)
-{
-	struct bfd_info *bfd_info;
-
-	if (!peer->bfd_info)
-		return;
-	bfd_info = (struct bfd_info *)peer->bfd_info;
+	/* Create new session and assign callback. */
+	p->bfd_config->session = bfd_sess_new(bfd_session_status_update, p);
+	bgp_peer_bfd_reset(p);
 
-	/* Check if the peer has been registered with BFD*/
-	if (!CHECK_FLAG(bfd_info->flags, BFD_FLAG_BFD_REG))
-		return;
+	/* Configure session with basic BGP peer data. */
+	if (p->su.sa.sa_family == AF_INET)
+		bfd_sess_set_ipv4_addrs(p->bfd_config->session,
+					p->su_local ? &p->su_local->sin.sin_addr
+						    : NULL,
+					&p->su.sin.sin_addr);
+	else
+		bfd_sess_set_ipv6_addrs(
+			p->bfd_config->session,
+			p->su_local ? &p->su_local->sin6.sin6_addr : NULL,
+			&p->su.sin6.sin6_addr);
 
-	bgp_bfd_peer_sendmsg(peer, ZEBRA_BFD_DEST_UPDATE);
-}
+	bfd_sess_set_vrf(p->bfd_config->session, p->bgp->vrf_id);
+	bfd_sess_set_hop_count(p->bfd_config->session,
+			       PEER_IS_MULTIHOP(p) ? p->ttl : 1);
 
-/**
- * bgp_bfd_reset_peer - reinitialise bfd
- * ensures that bfd state machine is restarted
- * to be synced with remote bfd
- */
-void bgp_bfd_reset_peer(struct peer *peer)
-{
-	if (!peer->bfd_info)
-		return;
+	if (p->nexthop.ifp)
+		bfd_sess_set_interface(p->bfd_config->session,
+				       p->nexthop.ifp->name);
 
-	bgp_bfd_peer_sendmsg(peer, ZEBRA_BFD_DEST_REGISTER);
+	bfd_sess_enable(p->bfd_config->session, true);
+	bgp_peer_config_apply(p, p->group);
```

Comment:
Should we remove this call from here and add it to `neighbor_bfd_cmd`.
This way it won't be called twice for all other commands.

Reply:
I think this is a non-issue: those functions will either do nothing (no changes) or will call `bfd_sess_install`.

If the `bfd_sess_install` is called multiple times in the same "event" (e.g. command or some other interaction) it will only send one update to `bfdd`:

```c
void bfd_sess_install(struct bfd_session_params *bsp)
{
	/* Don't attempt to install/update when disabled. */
	if (!bsp->enabled)
		return;

	bsp->lastev = BSE_INSTALL;
	thread_add_event(bsglobal.tm, _bfd_sess_send, bsp, 0, &bsp->installev);
}
```

It uses `bsp->installev` to track if there are multiple alterations in the same interaction and will only process the event once.

What do you think?

Reply:
If we use `neighbor <A.B.C.D|X:X::X:X|WORD> bfd (2-255) (50-60000) (50-60000)`, we'll first call `bgp_peer_configure_bfd(peer, true);` which will install the session with the default parameters.
Then it'll set the user-supplied parameters and call `bgp_peer_config_apply` again to update the session.
If we remove `bgp_peer_config_apply` call from `bgp_peer_configure_bfd`, the session will be  initially installed with needed parameters.

Also

Reply:
I see, I'll change it to have the all callers to call `apply` then. Nice catch, thanks!

Reply:
Ahh, now I get it. It won't actually install the session, because the event won't be processed. Then yes, it's not an issue.